### ISSUE TITLE
Nonterminating Sygus Problem (2)

### ISSUE THREAD

**Issue by BillHallahan** [CVC4/CVC4#3800]
**Describe the bug**

CVC4 does not seem to terminate on the below SyGuS problem (I've left it running for about 5 hours now), although a fairly small solution exists:
```
(set-logic ALL)
(define-fun safe-mod ((x Int) (y Int)) Int (mod x (+ 1 (abs y))))
(synth-fun args_0_refinement_0
	((r Int)) Bool
	((fv_B Bool) (B Bool) (fv_I Int) (I Int) (IConst Int))
	(
		(fv_B Bool ((Variable Bool) (Constant Bool) (= fv_I I) (= I I) (< fv_I I) (< I fv_I) (<= fv_I I) (<= I fv_I) (=> fv_B B) (=> B fv_B) (and fv_B B) (and B fv_B)))
		(B Bool ((Variable Bool) (Constant Bool) (= I I) (< I I) (<= I I) (=> B B) (and B B)))
		(fv_I Int (r (+ fv_I I) (+ I fv_I) (* fv_I I) (* I fv_I) (safe-mod fv_I IConst) 0))
		(I Int ((Variable Int) (+ I I) (+ I IConst) (- I I) (* I I) (safe-mod I IConst) 0))
		(IConst Int ((Constant Int)))
	)
)
(synth-fun ret_refinement_0
	((x0 Int) (r Int)) Bool 
	((fv_B Bool) (B Bool) (fv_I Int) (I Int) (IConst Int)) 
	(
		(fv_B Bool ((Variable Bool) (Constant Bool) (= fv_I I) (= I fv_I) (< fv_I I) (< I fv_I) (<= fv_I I) (<= I fv_I) (=> fv_B B) (=> B fv_B) (and fv_B B) (and B fv_B)))
		(B Bool ((Variable Bool) (Constant Bool) (= I I) (< I I) (<= I I) (=> B B) (and B B)))
		(fv_I Int (r (+ fv_I I) (+ I fv_I) (* fv_I I) (* I fv_I) (safe-mod fv_I IConst) 0))
		(I Int ((Variable Int) (+ I I) (- I I) (* I I) 0 4))
		(IConst Int ((Constant Int)))
	)
)
(constraint (=> (and (args_0_refinement_0 2)) (and (ret_refinement_0 2 8))))
(constraint (not (=> (and (args_0_refinement_0 2)) (and (ret_refinement_0 2 (- 1))))))
(constraint (and (and (ret_refinement_0 (- 2) (- 8)))))
(constraint (and (and (args_0_refinement_0 (- 2)))))
(constraint (and (and (ret_refinement_0 0 0))))
(constraint (and (and (args_0_refinement_0 0))))
(constraint (=> (and (args_0_refinement_0 0)) (and (ret_refinement_0 0 0))))
(constraint (not (=> (and (args_0_refinement_0 0)) (and (ret_refinement_0 0 (- 1))))))
(check-synth)
```
I apologize for the size of the problem: I tried to cut it down more, but was having trouble finding an even slightly smaller problem that doesn't terminate quickly.  For example, dropping `0` from _either_ `fv_I` or `I` in the grammar for `ret_refinement_0` allows CVC4 to solve the problem in under a second (and eliminate many other single terminals also allows the problem to be solved quickly.)  An example of a solution that's typically found is:
```
(define-fun args_0_refinement_0 ((r Int)) Bool (= 0 0))
(define-fun ret_refinement_0 ((x0 Int) (r Int)) Bool (= r (* x0 4)))
```

Should I expect CVC4 to be searching for over 5 hours on this problem?  Given the size of the gramamr, I'm not 100% sure that somehow eliminating single terminals isn't making the problem substantially easier, but it seems very suspicious to me?
 
**Command line arguments**: time cvc4 test2.sl --lang=sygus2
**CVC4 version/commit**:
This is CVC4 version 1.8-prerelease [git master 5489ef01]
compiled with GCC version 4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.33.17)
on Feb 20 2020 21:29:19

**Operating system**: MacOS

**Comment by ajreynol** [comment]
There is an issue with one of our heuristics, `--no-sygus-sym-break-pbe` solves both this problem and the one on #3789.  Will investigate this.

**Comment by BillHallahan** [comment]
Great, thanks for the tip about `--no-sygus-sym-break-pbe`.  I'll use that for now.

**Comment by ajreynol** [comment]
Ok, just FYI, disabling this option generally will make CVC4 significantly slower on PBE problems. I'll hopefully have a fix for this soon.

**Comment by ajreynol** [comment]
On closer inspection, `--no-sygus-sym-break-pbe` was a red herring.  The non-termination seems to be related to nonlinear arithmetic.  Generally, using grammars that admit non-linear terms can lead to the behavior you are seeing.  I will look into this a bit more to see if I can improve the combination of non-linear arithmetic and SyGuS, but since non-linear arithmetic is undecidable, this kind of behavior can be expected.

**Comment by BillHallahan** [comment]
Thanks Andy!  For many of the problems I care about, I think giving only `(* IConst I)` in the grammar (as opposed to the current `(* I I)` would be sufficient.  Would this change (and also similar changes to the other occurences of `*` in the grammar) be sufficient to take the grammar out of non-linear arithmetic, or might I still run into issues?

In https://github.com/CVC4/CVC4/issues/3580 you suggest a similar change for mod, to avoid non-linear arithmetic.  But it's unclear to me how `(Constant Int)` is being handled?

**Comment by ajreynol** [comment]
Yes, removing `(* I I)` should resolve these issues.  Your use of `safe-mod` looks right to me. 

Handling `(Constant Int)` is an open research problem :)  We are still working on improving our support for these kinds of grammars.  I am planning look into #3802 now.

**Comment by BillHallahan** [comment]
Got it, thanks!
